stdlib/public/core/StringUTF16View.swift

```diff
@@ -239,19 +245,16 @@ extension String.UTF16View: BidirectionalCollection {
   /// - Parameter position: A valid index of the view. `position` must be
   ///   less than the view's end index.
   @inlinable @inline(__always)
-  public subscript(i: Index) -> UTF16.CodeUnit {
-    String(_guts)._boundsCheck(i)
+  public subscript(idx: Index) -> UTF16.CodeUnit {
+    String(_guts)._boundsCheck(idx)
 
     if _fastPath(_guts.isFastUTF8) {
       let scalar = _guts.fastUTF8Scalar(
-        startingAt: _guts.scalarAlign(i)._encodedOffset)
-      if scalar.value <= 0xFFFF {
-        return UInt16(truncatingIfNeeded: scalar.value)
-      }
-      return scalar.utf16[i.transcodedOffset]
+        startingAt: _guts.scalarAlign(idx)._encodedOffset)
+      return scalar.utf16[idx.transcodedOffset]
```

Comment:
What would that information give us? We could also add that in the future if we want.

Regardless, we want a dedicated bit. It's more efficient and smaller code size than OR-ing up multiple bits (let along multiple branches), and this check happens *everywhere*.

Transcoded indices have scalar-aligned offsets, which we already detect (and set when we move on to the next scalar). We might want to clarify the semantics of the aligned bit to mean that there could still be a transcoded offset, i.e. it means the encoded offset was aligned. Then transcoded views would also always set the bit, rather than only setting it when we advance to the next scalar. It would also clean up a branch in transcoded code unit views where we check if transcoded is zero or alignment bit is set. On the flip side, anything that needs a aligned and non-transcoded index (e.g. `distance`) would have to check the alignment bit *and* transcoded offset.

Reply:
BTW, I added more comments and a table in the new update, such as https://github.com/apple/swift/pull/23834/files#diff-8ec9c169d4bc3c678838b4067481979eR211